### ISSUE TITLE
How to use getContentStream from an Isolate? - Background isolates do not support setMessageHandler()

### ISSUE THREAD

**Issue by tomekit** [talesbarreto/uri_content#10]
I am using this package to convert Android's ContentUri to Dart Stream which I can then subsequently process and e.g. upload file contents to the remote server.

This code works great
```dart
final uriContent = UriContent();
final stream = uriContent.getContentStream(uri);
```

however it doesn't work from a background isolate throwing: `Unsupported operation: Background isolates do not support setMessageHandler(). Messages from the host platform always go to the root isolate.`

Even though Flutter implemented Platform channel execution from another isolate: https://github.com/flutter/flutter/issues/13937 it seems Flutter doesn't allow to two-way communication fully: https://github.com/flutter/flutter/issues/119207

I am not sure if there is any feasible workaround that could be used (e.g. using some stream buffer (e.g. couple MBs) on the Android's native side and then polling for new stream chunks from Flutter).

Ideally Flutter framework shall fully support Platform channels from the background, I've still thought it's worth mentioning it here.

**Comment by talesbarreto** [comment]
Interesting!

If you are experiencing performance issues reading `content uri`, this may be related to https://github.com/talesbarreto/uri_content/issues/9.🤔. This synchronization is not difficult to implement but, unfortunately, I'm quite busy this month 😞.

**Comment by tomekit** [comment]
> If you are experiencing performance issues reading content uri

As such I am not facing performance issues with Content URI, the reason I am using background isolate is because logic around file handling (transformation / encryption / upload) in my app is executed from within the Isolate, so as a side effect I am also calling: `getContentStream` from within the isolate.

I can't call the `getContentStream` outside isolate (to avoid above issue) and then pass stream inside isolate, since as such only primitive/serializable types can be passed to Isolate.

**Comment by talesbarreto** [comment]
In order to fix issue #9, I changed how the data is read on Android. Although it is not related, I kept in mind this issue while writing a new solution and I believe it is possible to use UriContent from another isolate now.

Unfortunately, I am not able to test this new solution property or the problem described in this issue due to my currently tight agenda. So I released a new `Release Candidate` with those changes. I will eventually release it as a final version. 

@tomekit, could you test the version `2.2.0-rc.1` and check if this problem persists? 

https://pub.dev/packages/uri_content/versions/2.2.0-rc.1

**Comment by tomekit** [comment]
That's amazing, I am busy this week, but I will very likely have chance to try it out next week. Thank you.

**Comment by tomekit** [comment]
I've tried: `2.2.0-rc.1`, first impression it works from within the isolate. That's great.

I've integrated it in my app and then started to play with it, even though I've experienced couple times somewhat unpredictable "read speeds" I can't really provide any example that would allow to reproduce this issue. I certainly need to connect to higher bandwidth network / LAN sever to confirm that there isn't any speed limiting factor.

One issue that I've found so far is that if I start single file upload and then start another one (in my case it would be started in another isolate), then existing stream that was processed fails with: "Request not found" in UriContentPlugin.kt:152 and the new one that was just about to start doesn't seem to return even single byte, as Dio complains that: `0 bytes written but expected 173905088`.

I've also received an error: "Can't start request with Id 0 because it already exists", that's UriContentPlugin.kt:139, however I am not exactly sure how I've managed to get to that point.

If I had more time I would probably try to debug `UriContentPlugin.kt` a little to understand better why these issues occur, but I am running low on time.

If by any chance you have time to address the concurrency issue (that is uploading more than one file in a single moment), then I would happily continue my testing next week.

Thank you for your work so far !

![Screenshot_20240607_225630](https://github.com/talesbarreto/uri_content/assets/7759113/858e33ce-315c-4b55-80de-dde9d447347d)
![Screenshot_20240607_223205](https://github.com/talesbarreto/uri_content/assets/7759113/64ad5358-9c08-4da2-970c-6b32773e5803)

**Comment by talesbarreto** [comment]
> I've tried: `2.2.0-rc.1`, first impression it works from within the isolate. That's great.
> 
> I've integrated it in my app and then started to play with it, even though I've experienced couple times somewhat unpredictable "read speeds" I can't really provide any example that would allow to reproduce this issue. I certainly need to connect to higher bandwidth network / LAN sever to confirm that there isn't any speed limiting factor.
> 
> One issue that I've found so far is that if I start single file upload and then start another one (in my case it would be started in another isolate), then existing stream that was processed fails with: "Request not found" in UriContentPlugin.kt:152 and the new one that was just about to start doesn't seem to return even single byte, as Dio complains that: `0 bytes written but expected 173905088`.
> 
> I've also received an error: "Can't start request with Id 0 because it already exists", that's UriContentPlugin.kt:139, however I am not exactly sure how I've managed to get to that point.
> 
> If I had more time I would probably try to debug `UriContentPlugin.kt` a little to understand better why these issues occur, but I am running low on time.
> 
> If by any chance you have time to address the concurrency issue (that is uploading more than one file in a single moment), then I would happily carry on with testing.
> 
> Thank you for your work so far !

Wow 😱! That is very weird. Maybe I misunderstood how isolates interact with platform channels.

In the package example, we have several photos loading simultaneously. Everything is working there. I was expecting reading from an isolate would only interfere with the Dart side. 

On the Kotlin side, we have concurrence being managed using the coroutine library. The execution is controlled by mutexes that handle multithreading. Things should work even if they are accessed by different threads 🤔 .

I can not promise (seriously, things are crazy here 🙈) but I think I will have some spare time next weekend. I will implement an example that uses isolates to read those data and try debugging those issues.

Thank you so much for your help and your time.

**Comment by talesbarreto** [comment]
![](https://media1.tenor.com/m/1VWjeiT8gT0AAAAC/stupid-sport.gif)

Happily, the bug was obvious and easy to fix. To be able to read files simultaneously from Android, each reading needs a unique ID. I was using a singleton integer and increasing it each time that I needed a new ID. Obviously, that would never work in multiple isolates. 

I got rid of any singletons. The IDs are now generated by random numbers. 

I implemented a test using isolates in the example https://github.com/talesbarreto/uri_content/blob/main/example/lib/android_content/data_fetcher.dart#L26. It worked on my tests.

### speed reading

To address that `read speeds` issue that you reported, I increased the data chunk size that is read on each interaction between Dart/Kotlin.

In this new reading strategy, there is a considerable overhead to setup the reading of each chunk. Ideally, we should continuously send data from Kotlin to Dart (as it was before) and Dart should inform the native code when the stream should be paused.

As you pointed out in this issue, we can not directly send data from Kotlin to Dart when the code is not being executed from the main isolate. To overcome this, The Dart side requests each chunk individually, one by one after the stream consumes the previous one. So, yeap... we have more overhead now. Increasing the chunk size (that was ruinously low) should fix this problem. 

You can try different values by setting the `bufferSize`

https://github.com/talesbarreto/uri_content/blob/e126881eea0fe7da82f73fb7850cfa11e4a589fe/lib/src/uri_content.dart#L103-L111


@tomekit. I released the fix in the version `2.2.0 rc.2` https://pub.dev/packages/uri_content/versions/2.2.0-rc.2. If you are able to test when you have time, that would be awesome.

**Comment by talesbarreto** [comment]
Everything worked just fine in my tests. The version `2.2.0` is now available. If you find any bugs or unexpected behaviours, please, feel free to open a new issue.
